test/test_python_dispatch.py

```diff
@@ -0,0 +1,257 @@
+import torch
+from torch.testing._internal.common_utils import TestCase, run_tests
+from torch.utils._pytree import tree_map
+
+from typing import Iterator, List
+import logging
+import contextlib
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
```

Comment:
Just want to confirm two questions around `__torch_dispatch__` best practices, assuming the two common use cases of:
- "is a" (where the subclass uses itself as the actual tensor, and doesn't contain any tensor members)
- "has a" (where the subclass just treats itself as a meta tensor, and has at least one tensor member that it calls ops from `__torch_dispatch__` on)

(a) a context manager like `no_dispatch()` isn't needed in the vanilla "has a" case, and is only important in the "is a" case. Assuming that the member tensors are just ordinary tensors and not themselves subclasses, we don't have to worry about recursively hitting the python key again.

(b) Depending on if/how we decide to expose a redispatching API, it seems like that could probably replace the need for this context manager (at least in the vanilla "is a" case)? Since `DispatchKey::Python` has lower precedence than all other non-backend keys, we'd expect to go through all the other behavioral keys before hitting torch_dispatch, so the common case would probably be to just directly dispatch to the backend kernel.

I.e. instead of
```
with no_dispatch():
            output = func(*args)
```
We have something like
```
output = func.redispatch(*args) # dispatches below Python key, to backend kernel
```

Reply:
Yes, agreed on both counts.